### PR TITLE
Add argument to mstore

### PR DESCRIPTION
**Pull request by leonardoalt**
This decreases the number of used rows in the trace by 15% in memory heavy cases (like my Merkle tree example).

The `addr` register can be fully removed once `mload` is also modified similarly.

### CODE REVIEW

File: riscv/src/compiler.rs [hunk 1]
```diff
@@ -1242,10 +1246,7 @@ fn process_instruction(instr: &str, args: &[Argument], coprocessors: &CoProcesso
         }
         "sw" => {
             let (r1, r2, off) = rro(args);
-            vec![
-                format!("addr <== wrap({r2} + {off});"),
-                format!("mstore {r1};"),
-            ]
+            vec![format!("mstore {r2} + {off}, {r1};")]
```
[the remark this answers is not in the document]
Comment (georgwiese) [reply riscv/src/compiler.rs:1249]: It's done as part of the `mstore` instruction itself now.

**Comment by leonardoalt** [comment]
> Why did we use `addr` in the first place?

Don't remember, but this is from the very first version so, there was probably some constraint we solved on the way

File: riscv/src/compiler.rs [hunk 2]
```diff
@@ -1264,7 +1265,7 @@ fn process_instruction(instr: &str, args: &[Argument], coprocessors: &CoProcesso
                 format!("tmp3 <== and({rs}, 0xffff);"),
                 "tmp3 <== shl(tmp3, 8 * tmp2);".to_string(),
                 "tmp1 <== or(tmp1, tmp3);".to_string(),
-                "mstore tmp1;".to_string(),
+                "mstore addr, tmp1;".to_string(),
```
Comment (leonardoalt) [inline riscv/src/compiler.rs:1268]: This case and the two below still require the address to be saved somewhere at first. When we remove `addr` completely we can use something else. If there are no registers available we can rename `addr` to `tmp4`.

File: riscv/src/compiler.rs [hunk 3]
```diff
@@ -501,7 +501,11 @@ fn preamble(coprocessors: &CoProcessors) -> String {
 
     // ============== memory instructions ==============
 
-    instr mstore X { { addr, STEP, X } is m_is_write { m_addr, m_step, m_value } }
+    instr mstore Y, Z {
```
Comment (chriseth) [inline riscv/src/compiler.rs:504]: Can you document the allowed range for Y?

File: riscv/src/compiler.rs (hunk 2 above)
Comment (chriseth) [reply riscv/src/compiler.rs:1268]: I think we can just use `{rd} + {off}` here.

Comment (chriseth) [reply riscv/src/compiler.rs:1268]: ok, not exactly.

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
> Why did we use `addr` in the first place?

because we did not have multi-argument instructions

**Pull request merged**